corehq/apps/locations/queryutil.py

```diff
@@ -0,0 +1,158 @@
+    hence the name and slightly verbose implementation.
+    """
+
+    def __init__(self, mptt_set, timing_context):
+        self._mptt_set = mptt_set
+        if isinstance(timing_context, ComparedQuerySet):
+            timing_context = timing_context._timing.clone()
+        self._timing = timing_context
+
+    def __iter__(self):
+        with _commit_timing(self):
+            with self._timing("mptt") as timer1:
+                for item in self._mptt_set:
+                    with timer1.pause():
+                        yield item
+
+    def __len__(self):
+        with _commit_timing(self):
+            with self._timing("mptt"):
+                return len(self._mptt_set)
+
+    def get(self, *args, **kw):
+        with _commit_timing(self):
+            with self._timing("mptt"):
+                return self._mptt_set.get(*args, **kw)
+
+    def exists(self, *args, **kw):
+        with _commit_timing(self):
+            with self._timing("mptt"):
+                return self._mptt_set.exists(*args, **kw)
+
+    def accessible_to_user(self, domain, user):
+        # mostly copied from models.py:LocationQueriesMixin
+
+        if user.has_permission(domain, 'access_all_locations'):
+            return self.filter(domain=domain)
+
+        assigned_location_ids = user.get_location_ids(domain)
+        if not assigned_location_ids:
+            return self.none()  # No locations are assigned to this user
+
+        from .models import SQLLocation
+        ids_query = SQLLocation.objects.get_locations_and_children(assigned_location_ids)
+        assert isinstance(ids_query, ComparedQuerySet), ids_query
+        result = ComparedQuerySet(
+            self._mptt_set.filter(id__in=ids_query._mptt_set),
+            TimingContext("accessible_to_user"),
+        )
+        result._timing += ids_query._timing
+        return result
+
+
+def _make_method(name):
+    def method(self, *args, **kw):
+        return ComparedQuerySet(
+            getattr(self._mptt_set, name)(*args, **kw),
+            self,
+        )
+    method.__name__ = str(name)  # unicode_literals: must be bytes on PY2
+    return method
+
+
+for name in [
+    "annotate",
+    "filter",
+    "none",
+    "order_by",
+    "values",
+    "values_list",
+    # from subclasses of QuerySet
+    "location_ids",
+]:
+    setattr(ComparedQuerySet, name, _make_method(name))
+
+
+@contextmanager
+def _commit_timing(queryset):
+    # only send to datadog on initial query evaluation
+    commit = queryset._mptt_set._result_cache is None
+    try:
+        yield
+    finally:
+        if commit and queryset._mptt_set._result_cache is not None:
+            timing = queryset._timing
+            for key in timing.timers:
+                bucket = bucket_value(timing.duration(key), TIME_BUCKETS, "s")
+                datadog_counter(
+                    'commcare.locations.%s.%s.count' % (timing.name, key),
+                    tags=['duration:%s' % bucket],
```

Comment:
what are your thoughts on adding domain as a tag?

Reply:
Good idea, although I'll likely only tag the important ones as is done [here](https://github.com/dimagi/commcare-hq/blob/66565ba5fa16764f9072b3552f16c2de8f7f31c5/corehq/ex-submodules/casexml/apps/phone/restore.py#L719-L720). Tagging every domain is not great because it balloons the [number of custom metrics in datadog](https://help.datadoghq.com/hc/en-us/articles/204271775-What-is-a-custom-metric-and-what-is-the-limit-on-the-number-of-custom-metrics-I-can-have-).

Reply:
@emord when I went to implement this I realized that most of the time `ComparedQuerySet` is too low-level and has no knowledge of the domain (exception: `accessible_to_user(domain, user)`). I don't think there is an easy way to get the domain without pushing the timing logic up to where the specific queries are being composed. That would require an entirely different approach from what's being done in this PR, so I'm not going to do it. Let me know if you think I should reconsider.